saleor/graphql/discount/mutations/promotion/validators.py

```diff
@@ -57,6 +58,31 @@ def clean_promotion_rule(
     return cleaned_input
 
 
+def _clean_gifts(cleaned_input, instance, errors, error_class, index):
+    """Check if assigned gifts extends the gift limit."""
+    if "gifts" not in cleaned_input:
+        return
+
+    gifts = {gift.id for gift in cleaned_input["gifts"]}
+    if instance:
+        # instance is provided only for PromotionRuleUpdate mutation
+        # so the gifts will be fetched once
+        gifts |= {gift.id for gift in instance.gifts.all()}
```

Comment:
When you update `gifts` field with `PromotionRuleUpdate` mutation current gifts are replaced with the ones from the input. Here you combine both sets what is wrong.